### ISSUE TITLE
Investigate moving (some features) to plugins

### ISSUE THREAD

**Issue by Ornias1993** [jailmanager/jailman#196]
# Issue Template
## Description

Currently we are already pretty close to iocage plugins, but we add persistence, interlinking and, for some jails, improved performance.

It might be interesting moving some jails to customized plugins, because they don't require intervention from host during setup.

**Comment by Ornias1993** [comment]
@thorsteneb This might also be a relatively easy fix for Plex hardware transcoding...
We can create a plugin: 
- Based on de latest FreeBSD PKG
- Including the required changes for hardware transcoding

**Comment by thorsteneb** [comment]
Yeah I've been thinking along those lines. I'm not too enamored with the fact that whatever is in options.json is not exposed via UI. And, still, this would be pretty clean. Just needs one tunable, and no real choices on the user part. As far as I know, plugins run in jail context though during install, so shenanigans like mounting a config dataset in are right out. Pros and cons to this approach.

**Comment by Ornias1993** [comment]
True.
I was thinking of a middle ground:
- The plugins would primarily just be internal for now (install a plugin instead of an empty jail)
- They would be pretty basic (Pkg, repo, release, overlay-file-structure and sometimes(!) a post-install script)

It basically would just shuffle a few things:
- The PKG list in config.yml would move to the plugin config
- The repo would be easily set able
- The "includes" folder would be changed into an "overlay" folder for the plugin

All network settings and mount options should still be available for the plugin and since about a year we can override the name of the jail created from a plugin

I was thinking about this due to the plex issue, but also the fact we are now doing things that are already included in iocage-plugins...

**Comment by Ornias1993** [comment]
Also, `service get` and `service set` are interesting options!
An example jail that uses those:
https://github.com/freenas/iocage-plugin-zoneminder

**Comment by Ornias1993** [comment]
@thorsteneb If we can get the following request to pass, this basically makes plugins our future:
https://github.com/iocage/iocage/issues/1191

If we can just get those /config folders mounted into the jails and have our persistent storage ready after install... We could significantly cut our code....

**Comment by Ornias1993** [comment]
I also went ahead an made a mock-up for a plex hardware transcoding compatible plugin:
https://github.com/jailmanager/iocage-plugins/blob/master/plex.json
https://github.com/jailmanager/iocage-plugin-plex

**Comment by jsegaert** [comment]
> I'm not too enamored with the fact that whatever is in options.json is not exposed via UI.

Same here.  I added a feature request for this a long time ago but nobody ever looked at it.  I now re-entered it as a "community suggestion" in JIRA, so that the community can vote on it.  Please take the time to vote on it.  😊 

https://jira.ixsystems.com/browse/NAS-106792

**Comment by Ornias1993** [comment]
Even when we use plugins, many will still be incompatible with normal UI-based installations I think...
Let met explain:
- We will push additional settings to the installed plugin (something plugins natively do not allow), defaults are not always an option (such as database connection details)
- We fetch details from other jails (For example Nextcloud fetches the IP for the listed mariadb jail)
- We also setup other jails accordingly (such as adding an extra database, by iocage exec on the database jail)

Assuming we get persistent storage for plugins into iocage, we could just write a .env file from jailman before installing the plugin and fetch it from post install.

Currently we do:
Create Jail -> send data to jail -> modify other jails -> post_install

With plugins this would become:
Modify other jails -> send data to jail-storage -> create jail form plugin -> post_install

Nice side effect from this, is that more errors will be detected before we start jail creation.

Why move to plugins?
- Standardized formatting makes Jailman more accessible
- Setting repo, overlay-filecopy etc is handled by iocage
- We can set devfs_rules
- Plugin maintainers can be (easily) set to be overlord of their own repository (instead of global write)

Simply put: I think we are doing a lot of things iocage has already figured out for us, moving to plugins would help us focus on what we really want to do: Write great jail management.

**Comment by Ornias1993** [comment]
Thanks for all the feedback, I formulated a proposal based on this and #171
Hence closing.
